### PR TITLE
rework chdir handling use mktemp instead of '/tmp'

### PR DESCRIPTION
**Pull request by ttyS4**
There are multiple changes here. It displays my current understanding of how these modules do chdir.

The current behavior is:

1. NCD::CLI enters `/tmp`
2. NCD::ComponentProxy enters `/tmp`
3. a component is executed
4. NCD::ComponentProxy restores to `/tmp` (which NCD::CLI set up)
5. then back to step 2.

New proposed behavior after this change:

Instead of `/tmp` a `/tmp/ncd-components-XXXXXXXX` (x's replaced with random alphanumerics) is used. (This is good because `/tmp` is a hostile environment being a world-writable directory.)
Step 1. removed. Step 4. kept because we need to leave the tempdir to let the module to automatically clean it up.

If a module changes the current directory during its execution Step 2. still changes back to the previous.

1. NCD::ComponentProxy enters `/tmp/ncd-components-XXXXXXXX`.
2. a component is executed
3. NCD::ComponentProxy restores to whatever the ncm-ncd was invoked from
4. then back to step 1.

### CODE REVIEW

File: src/main/perl/CLI.pm [hunk 1]
```diff
@@ -505,10 +504,6 @@ sub action
         chroot($self->option("chroot")) or die "Unable to chroot to ", $self->option("chroot");
     }
 
-    if(! chdir($NCD_CHDIR)) {
```
Comment (stdweird) [inline src/main/perl/CLI.pm]: why did you remove this?

File: src/main/perl/ComponentProxy.pm [hunk 2]
```diff
@@ -21,7 +22,7 @@ my $ec = LC::Exception::Context->new->will_store_all;
 
 use Readonly;
 Readonly my $COMPONENTS_PROFILE_PATH => '/software/components';
-Readonly my $RUN_FROM => '/tmp';
+Readonly my $RUN_FROM => File::Temp->newdir('/tmp/ncd-components-XXXXXXXX');
```
Comment (stdweird) [inline src/main/perl/ComponentProxy.pm]: can you add `CLEANUP => 1` (i'm aware it's the default, but it's better if it is made explicit)
@ttyS4 @ned21 you are aware that this removes EL5 support? or requires a recent `FILE::Temp` version?

Comment (stdweird) [reply src/main/perl/ComponentProxy.pm]: see also `CAF::Path` https://github.com/quattor/CAF/blob/master/src/main/perl/Path.pm#L383.
it support temp directory creation and error handling

File: src/main/perl/CLI.pm (hunk 1 above)
Comment (stdweird) [reply src/main/perl/CLI.pm]: in particular, does chroot still work? the tempdir is created during import, does it also exist in the chroot environment?

Comment (ttyS4) [reply src/main/perl/CLI.pm]: As summarized above it seemed redundant to me to do the chdir in `NCD::CLI` because it happens before each module execution. Is it needed?

For chroot - I think you are right and I need to make adjustments to the code to create the directory in the chroot.

File: src/main/perl/ComponentProxy.pm (hunk 2 above)
Comment (ttyS4) [reply src/main/perl/ComponentProxy.pm]: Sure, i can add `CLEANUP` to the call.
For RH5, I think it should just work fine. I can test tomorrow.

So you want me to rework to use CAF::Path (it uses File::Temp itself) or it is optional?

Comment (stdweird) [reply src/main/perl/ComponentProxy.pm]: no, i'm pointing out that in CAF::Path newdir isn't used because of EL5 support. correct ref is https://github.com/quattor/CAF/blob/master/src/main/perl/Path.pm#L409
but feel free to use CAF::Path. that way we can add unittests more easily.

File: src/main/perl/CLI.pm (hunk 1 above)
Comment (stdweird) [reply src/main/perl/CLI.pm]: main reason is to do this asap in the commandflow. but i'm not sure it has that much advatange.
as along as you define and the create the tempdir only once.

**Comment by jrha** [comment]
@stdweird what state is this in?

**Comment by stdweird** [comment]
@jrha it's not ready

**Comment by ttyS4** [comment]
I am going to get back with my updates by the end of the week.

**Comment by jrha** [comment]
@ttyS4 Ok thanks, I'd like to finalise the release soon.

**Comment by ttyS4** [comment]
I am experimenting with this: https://github.com/ttyS4/ncm-ncd/commit/d67cbb6d4eeca9e8c19dea454434320e188a5c82
however I got stuck and need some help.

I do not know how can I get the chroot setting and would need some guidance.

**Comment by jrha** [comment]
@ned21 I want to finalise the release this week, would you guys mind if we bumped this to 18.3?

**Comment by ttyS4** [comment]
@jrha #123 should replace this and is is ready to merge in.

**Comment by jrha** [comment]
Closing as replaced by #123.
